auth/user.go

```diff
@@ -216,7 +231,7 @@ func (user *userImpl) RevokedChannels(since uint64) RevokedChannels {
 	for roleName, history := range roleHistory {
 		if !user.RoleNames().Contains(roleName) {
 			for _, entry := range history.Entries {
-				if entry.StartSeq <= since && entry.EndSeq > since {
+				if entry.StartSeq <= checkSeq && entry.EndSeq > checkSeq || (entry.StartSeq <= triggeredBy && entry.EndSeq > triggeredBy) {
```

Comment:
I need to add parentheses to the first two.
We do need this so that we can obtain any revocations that happened AT the triggeredBy. At this point we don't yet know whether this is the resumption of an interrupted feed or whether it's just a since with a triggeredBy.
Once this function returns we will have the seq at which a given channel is revoked at. This allows us to do that calculation where we check whether revocationSeq is equal to the triggeredBy so we know whether we need to resume revocations or check from 0.

Reply:
The triggeredBy -1 check is for the following scenario:

```
- User has role1
- Role1 has ch1, ch2
- doc1 in ch1 at seq 10
- doc2 in ch2 at seq 11
- client does a pull. Last seq 11, has both docs.
- ch1 revoked from role at seq 20
- ch2 revoked from role at seq 30
- At this point the invalSeq on the role is at 20 as we keep the oldest.
- Mutate doc1 seq 40
- Mutate doc2 seq 50
```

We start a revocation with say since 11. This will find that ch1 and ch2 are revoked and start sending, however, after sending the first the replication is interrupted and we have a since of 20:40.

If we just use '20' as triggered by it won't flag that ch1 and ch2 are revoked because RevokedChannels will say 'well at 20 you didn't have any channels so we're good'. So we need -1 for that.

-----

The triggeredBy check is so we can find out if the revocation was interrupted or not. At the point where we're checking for RevokedChannels we don't know whether we have an interrupted replication or just a 'normal' replication which happens to have a triggeredBy in, perhaps the last sequence it got was a revocation. 

```
- Add user channel "A" 			seq3
- Add doc in "A"				seq4
- getChanges since 0			gets doc - lastSeq 4

- Remove channel "A" from user 	seq5
- Mutate doc 					seq6
- getChanges since 4			revokes doc - lastSeq 5:6

- Add user channel "A"			seq7
- getChanges since 5:6			grants - lastSeq 7:6

- Remove channel "A" from user 	seq8
- Mutate doc 					seq9
- getChanges since 7:6		
```	

That final stage SHOULD receive a revocation. However, if we don't check triggeredBy in RevokedChannels:
- checkSeq would be 6 (triggeredBy - 1)
- At 6 we didn't have access to the document so we don't need to revoke
We also need to check triggeredBy here too. The triggeredBy of 7 would say 'Okay yeah, you had access at 7: I'll return "A": 7 as a map entry'.

Reply:
(as an aside I double checked and 'AND' takes precedence over 'OR' which is why that evaluated fine